burstphoto/align.metal

```diff
@@ -416,3 +780,922 @@ kernel void compute_merge_weight(texture2d<float, access::read> texture_diff [[t
+    float const mismatch = 0.25f*(mismatch4[0] + mismatch4[1] + mismatch4[2] + mismatch4[3]);
+    
+    mismatch_texture.write(mismatch, gid);
+}
+
+
+kernel void normalize_mismatch(texture2d<float, access::read_write> mismatch_texture [[texture(0)]],
+                               constant float* mean_mismatch_buffer [[buffer(0)]],
+                               uint2 gid [[thread_position_in_grid]]) {
+    
+    // load args
+    float const mean_mismatch = mean_mismatch_buffer[0];
+    
+    float mismatch_norm = mismatch_texture.read(gid).r;
+    
+    // normalize that mean is set to 0.2
+    mismatch_norm /= (mean_mismatch*5.0f + 1e-12f);
+    
+    mismatch_norm = clamp(mismatch_norm, 0.0f, 1.0f);
+    
+    mismatch_texture.write(mismatch_norm, gid);
+}
+
+
+kernel void correct_hotpixels(texture2d<float, access::read> average_texture [[texture(0)]],
+                              texture2d<float, access::read> in_texture [[texture(1)]],
+                              texture2d<float, access::write> out_texture [[texture(2)]],
+                              constant float* mean_texture_buffer [[buffer(0)]],
+                              uint2 gid [[thread_position_in_grid]]) {
+       
+    int const x = gid.x+2;
+    int const y = gid.y+2;
+    
+    // load args
+    float mean_texture = 0.0f;
+    
+    if (x%2 == 0 & y%2 == 0) {
+        mean_texture = mean_texture_buffer[0];
+    }
+    if (x%2 == 1 & y%2 == 0) {
+        mean_texture = mean_texture_buffer[1];
+    }
+    if (x%2 == 0 & y%2 == 1) {
+        mean_texture = mean_texture_buffer[2];
+    }    
+    if (x%2 == 1 & y%2 == 1) {
+        mean_texture = mean_texture_buffer[3];
+    }
+        
+    // calculate sum of 8 pixels surrounding the potential hot pixel
+    float sum =   average_texture.read(uint2(x-2, y-2)).r;
+    sum      +=   average_texture.read(uint2(x+2, y-2)).r;
+    sum      +=   average_texture.read(uint2(x-2, y+2)).r;
+    sum      +=   average_texture.read(uint2(x+2, y+2)).r;
+    sum      += 2*average_texture.read(uint2(x-2, y+0)).r;
+    sum      += 2*average_texture.read(uint2(x+2, y+0)).r;
+    sum      += 2*average_texture.read(uint2(x+0, y-2)).r;
+    sum      += 2*average_texture.read(uint2(x+0, y+2)).r;
+    
+    float const pixel_value = average_texture.read(uint2(x, y)).r;
+    float const pixel_ratio = pixel_value/sum;
+    
+    // if hot pixel is detected
+    if (pixel_ratio >= 0.15f & pixel_value >= 2.0f*mean_texture) {
+        
+        float sum2 = in_texture.read(uint2(x-2, y+0)).r;
+        sum2      += in_texture.read(uint2(x+2, y+0)).r;
+        sum2      += in_texture.read(uint2(x+0, y-2)).r;
+        sum2      += in_texture.read(uint2(x+0, y+2)).r;
+        
+        float const weight = 10.0f*min(pixel_ratio-0.15f, 0.10f);
+        
+        out_texture.write(weight*0.25f*sum2 + (1.0f-weight)*pixel_value, uint2(x, y));
+    }
+}
+
+
+kernel void forward_fft(texture2d<float, access::read> in_texture [[texture(0)]],
```

Comment:
The new method seems to struggle with specular highlights and movement.

This is a crop of the original image area
![Orig](https://user-images.githubusercontent.com/48962821/213606068-7971949c-723a-461e-a672-0fe4912c9e86.jpg)

This I a crop with noise reduction set to 1:
![NR=1](https://user-images.githubusercontent.com/48962821/213606982-032a12d9-f186-4e59-ac34-0bc7b3f82f9f.jpg)

And this is it set to 25 to show the movement:
![NR=25](https://user-images.githubusercontent.com/48962821/213607007-a4df6e6e-5bbd-4cef-b75f-868bbf15ab22.jpg)

The problem happens even if NR=20. And did not happen with the previous method

(I will see about getting other raw files that produce these issues, I don't want to post the full images.)

Reply:
The behavior for specular highlights is currently a limitation of the frequency-based merging and also described in the original publication in the artifacts section:
https://people.csail.mit.edu/hasinoff/pubs/HasinoffEtAl16-hdrplus.pdf

However, I could imagine that it can be improved. In case you have a burst series with such artifacts you can share privately, I would be happy to have a look at it...

Reply:
I haven't been able to recreate another set of photos to do it.

However, the latest changes you made did help

![After](https://user-images.githubusercontent.com/48962821/214193417-288d5072-19bb-4070-8b73-55d4865f76c0.png)

However, I looked through the photos in the article, and it looks like they simply did not merge anything on the tiles with specular highlights (figure 8).

Reply:
The specular highlights can have very high contrast and sharpness and as a consequence lead to a very extreme frequency spectrum. When using differences in frequencies for calculating the merging weights, this extreme behavior results in merging weights near zero. That is a clear limitation of this approach. However, it occurs only with very high-contrast specular highlights.

Reply:
> The specular highlights can have very high contrast and sharpness and as a consequence lead to a very extreme frequency spectrum. When using differences in frequencies for calculating the merging weights, this extreme behavior results in merging weights near zero. That is a clear limitation of this approach. However, it occurs only with very high-contrast specular highlights. 

My point was more that in the paper it appears that their failure mode is a lack of merging rather than the presence of color artifacts like above.

I will see about finding which pictures from their dataset are used for figure 8, and see if the method as implemented here produces the color artifacts of the lack of merging.

Reply:
Ah ok I see. Actually I am also surprised that you got these color artifacts. The artifacts I am used to at specular highlights are more similar to the ones shown in the publication. But maybe your reference image has exactly this noise pattern at this position.

Reply:
It's actually not just that one spot. I've also found it to happen with the reflection of light sources in people's eyes, specifically over the pupil. But that one is easy to miss since the original highlight is so tiny.

Reply:
Looked into this some more, and found the original files used to create Fig 8. You'll need a google account to download them.

You can download the original raw burst files from: https://console.cloud.google.com/storage/browser/hdrplusdata/20171106/bursts/47L8_20150518_203851_306?pageState=%28%22StorageObjectListTable%22%3A%28%22f%22%3A%22%255B%255D%22%29%29&prefix=&forceOnObjectsSortingFiltering=false

And their final merged dng and processed jpeg from: https://console.cloud.google.com/storage/browser/hdrplusdata/20171106/results_20161014/47L8_20150518_203851_306;tab=objects?prefix=&forceOnObjectsSortingFiltering=false

I've created a side by side screenshot (implementation in this PR on the left and their output on the right):
![diff](https://user-images.githubusercontent.com/48962821/218274437-013c1544-2e4b-428d-baf1-d4a5874de669.jpg)

Ignoring the performance through the image, what's clear is that the paper's implementation does not produce the artifacts around the specular highlights, it instead looks to be doing nothing around them. Notice the noise squares around each rivet on the right image, and the green highlights that appear around the left-most rivets in the left image.
Similar issues can be seen in order regions of the scene where there are specular highlights.

I'm starting to go through the paper and see what their doing and then will go through the code to see how they match up and get an understanding of what's going on.

Reply:
Hi Alex,

Thank you for investigating the artifacts around specular highlights further. It is good to have an example of an image burst where the color artifacts appear. 
One thing to consider when comparing results: the Google implementation probably also applies spatial denoising (at least my understanding of their publication) while my implementation does only apply temporal denoising. In my opinion, there are very sophisticated spatial denoising algorithms out there (e.g. DxO Photolab Deep Prime XD) and therefore spatial denoising should be applied in the RAW processor of choice.
However, the differences you observed have probably another reason. To reduce artifacts at tile borders, the Google publication suggests to perform the merging with overlapping tiles using raised cosine window for blending. Now the question is: apply the cosine window to the tiles BEFORE Fourier transformation or AFTER the inverse Fourier transformation. I tried both and decided to apply it BEFORE. My decision was based on the observation that when applying it AFTERWARDS, I got the same issue than you found in the Google image (see screenshot LEFT). Screenshot RIGHT shows the current result. From the signal processing perspective, it makes sense to apply a so-called apodization function (here: cosine window) BEFORE the Fourier transform. Otherwise very strong side lobes of the frequency spectrum are generated and as a consequence, the difference between the reference tile and aligned tile becomes large and in turn the merging weight small. The consequence is strong noise in patches with high frequency content (e.g. specular highlights).
I will try to investigate the reason for the color artifacts and have not yet an explanation for that.

<img width="974" alt="Bildschirm­foto 2023-02-11 um 21 51 10" src="https://user-images.githubusercontent.com/117006531/218280660-09cd9b66-130c-4885-875d-141f6d8aaf90.png">

Reply:
I just downloaded the Google burst and can reproduce your findings. You are right, would be great to connect via some private chat option. Can fully understand that my answer may take some time to digest :-)

Reply:
@martin-marek, could you please create a Gitter for this project here: https://gitter.im/home/explore#createcommunity

I can create my own room, but if we're gonna discuss things for the project, I think it makes the most sense for you to create one officially for the project. It should only take a minute or two.

Reply:
Found a bug in the hot pixel correction that caused the color artifacts. Will push the changes to GitHub within the next days...
![Bildschirm­foto 2023-02-12 um 00 58 08](https://user-images.githubusercontent.com/117006531/218286124-f48fa9f7-a366-4026-87f4-10fb6a61a87d.png)

Reply:
Tried it with other cases and can confirm that last push with the bug fix solved it. 👍